File: server/service/author/author_service.go
package authorService

import (
	"context"
	"github.com/nnniyaz/blog/domain/author"
	"github.com/nnniyaz/blog/domain/base/uuid"
	"github.com/nnniyaz/blog/repo"
)

type AuthorService interface {
	Create(ctx context.Context, firstName, lastName, avatarUri string) error
	Update(ctx context.Context, id, firstName, lastName, avatarUri string) error
	Delete(ctx context.Context, id string) error
	Restore(ctx context.Context, id string) error
	FindById(ctx context.Context, id string) (*author.Author, error)
	FindAll(ctx context.Context) ([]*author.Author, int64, error)
}

type authorService struct {
	repo repo.Author
}

func NewAuthorService(repo repo.Author) AuthorService {
	return &authorService{repo: repo}
}

func (s *authorService) Create(ctx context.Context, firstName, lastName, avatarUri string) error {
	newAuthor, err := author.NewAuthor(firstName, lastName, avatarUri)
	if err != nil {
		return err
	}
	return s.repo.Create(ctx, newAuthor)
}

func (s *authorService) Update(ctx context.Context, id, firstName, lastName, avatarUri string) error {
	convertedId, err := uuid.UUIDFromString(id)
	if err != nil {
		return err
	}

	author, err := s.repo.FindById(ctx, convertedId)
	if err != nil {
		return err
	}

	err = author.Update(firstName, lastName, avatarUri)
	if err != nil {
		return err
	}
	return s.repo.Update(ctx, author)
}

func (s *authorService) Delete(ctx context.Context, id string) error {
	convertedId, err := uuid.UUIDFromString(id)
	if err != nil {
		return err
	}
	return s.repo.Delete(ctx, convertedId)
}

func (s *authorService) Restore(ctx context.Context, id string) error {
	convertedId, err := uuid.UUIDFromString(id)
	if err != nil {
		return err
	}
	return s.repo.Restore(ctx, convertedId)
}

func (s *authorService) FindById(ctx context.Context, id string) (*author.Author, error) {
	convertedId, err := uuid.UUIDFromString(id)
	if err != nil {
		return nil, err
	}
	return s.repo.FindById(ctx, convertedId)
}

func (s *authorService) FindAll(ctx context.Context) ([]*author.Author, int64, error) {
	return s.repo.FindAll(ctx)
}

File: crm/src/http/author/authorService.ts
import {AxiosResponse} from 'axios'

import {Author} from '../../domain/author/author'
import {MlString} from '../../domain/base/mlString/mlString'
import {$api, ApiRoutes, Request} from '../index'
import {ErrorResponse, SuccessResponse} from '../response/response'

type AuthorServiceGetAllOut = {
    count: number
    authors: Author[]
}

type AuthorServiceGetByIdIn = {
    id: string
}

type AuthorServiceCreateIn = {
    firstName: MlString
    lastName: MlString
    avatarUri: string
}

type AuthorServiceUpdateIn = {
    id: string
    firstName: MlString
    lastName: MlString
    avatarUri: string
}

type AuthorServiceDeleteIn = {
    id: string
}

type AuthorServiceRestoreIn = {
    id: string
}

export class AuthorService {
    static async getAll(request: Request<null>): Promise<AxiosResponse<SuccessResponse<AuthorServiceGetAllOut> | ErrorResponse>> {
        return $api(request.lang).post<SuccessResponse<AuthorServiceGetAllOut> | ErrorResponse>(ApiRoutes.AUTHOR_GET_ALL, null, {signal: request.controller?.signal})
    }

    static async getById(request: Request<AuthorServiceGetByIdIn>): Promise<AxiosResponse<SuccessResponse<Author> | ErrorResponse>> {
        return $api(request.lang).post<SuccessResponse<Author> | ErrorResponse>(ApiRoutes.AUTHOR_GET_BY_ID.replace(':id', request.body.id), null, {signal: request.controller?.signal})
    }

    static async create(request: Request<AuthorServiceCreateIn>): Promise<AxiosResponse<SuccessResponse<null> | ErrorResponse>> {
        return $api(request.lang).post<SuccessResponse<null> | ErrorResponse>(ApiRoutes.AUTHOR_CREATE, null, {signal: request.controller?.signal})
    }

    static async update(request: Request<AuthorServiceUpdateIn>): Promise<AxiosResponse<SuccessResponse<null> | ErrorResponse>> {
        return $api(request.lang).post<SuccessResponse<null> | ErrorResponse>(ApiRoutes.AUTHOR_UPDATE.replace(':id', request.body.id), null, {signal: request.controller?.signal})
    }

    static async delete(request: Request<AuthorServiceDeleteIn>): Promise<AxiosResponse<SuccessResponse<null> | ErrorResponse>> {
        return $api(request.lang).post<SuccessResponse<null> | ErrorResponse>(ApiRoutes.AUTHOR_DELETE.replace(':id', request.body.id), null, {signal: request.controller?.signal})
    }

    static async restore(request: Request<AuthorServiceRestoreIn>): Promise<AxiosResponse<SuccessResponse<null> | ErrorResponse>> {
        return $api(request.lang).post<SuccessResponse<null> | ErrorResponse>(ApiRoutes.AUTHOR_RESTORE.replace(':id', request.body.id), null, {signal: request.controller?.signal})
    }
}